Fix: Bedrock inference profiles for Nova Canvas image generation

timelfrink opened the pull request:
## Title

Fix: Bedrock inference profiles for Nova Canvas image generation

## Relevant issues

Fixes #14373

## Pre-Submission checklist

**Please complete all items before asking a LiteLLM maintainer to review your PR**

- [x] I have Added testing in the [`tests/litellm/`](https://github.com/BerriAI/litellm/tree/main/tests/litellm) directory, **Adding at least 1 test is a hard requirement** - [see details](https://docs.litellm.ai/docs/extras/contributing_code)
- [x] I have added a screenshot of my new test passing locally
- [x] My PR passes all unit tests on [`make test-unit`](https://docs.litellm.ai/docs/extras/contributing_code) - 43/44 tests passed, 1 failure due to missing boto3 dependency (expected in test environment)
- [x] My PR's scope is as isolated as possible, it only solves 1 specific problem

## Type

🐛 Bug Fix

## Changes

### Problem Being Solved

Bedrock Application Inference Profiles were not working with Amazon Nova Canvas image generation models, causing two critical errors:

1. **ARN Format Error**: When using ARN format inference profiles (`arn:aws:bedrock:region:account:application-inference-profile/profile-id`), users encountered:
   ```
   BedrockError: Unsupported model=arn:aws:bedrock:eu-west-1:000000000000:application-inference-profile/a0a0a0a0a0a0, passed in
   ```

2. **Environment Variable Error**: When using `model_id: os.environ/MODEL_ID_CANVAS` configuration, users encountered:
   ```
   BedrockError: {"message":"Malformed input request: #/imageGenerationConfig: extraneous key [model_id] is not permitted, please reformat your input and try again."}
   ```

### Root Cause Analysis

The issue was in the image generation handler's provider detection logic:

- **File**: `litellm/llms/bedrock/image/image_handler.py:236`
- **Problem**: Used naive `model.split(".")[0]` which fails for ARN formats
- **Secondary Issue**: `model_id` parameter wasn't filtered, causing API validation errors

### Implementation Details

#### 1. Enhanced Provider Detection (`image_handler.py`)

**Before**:
```python
provider = model.split(".")[0]
```

**After**:
```python
# Use the existing ARN-aware provider detection method
bedrock_provider = self.get_bedrock_invoke_provider(model)

if bedrock_provider == "amazon" or bedrock_provider == "nova":
    # Handle Amazon Nova Canvas models
    provider = "amazon"
elif bedrock_provider == "stability":
    provider = "stability"
else:
    # Fallback to original logic for backward compatibility
    provider = model.split(".")[0]
```

#### 2. Parameter Filtering (`amazon_nova_canvas_transformation.py`)

**Before**:
```python
image_generation_config = {**image_generation_config, **optional_params}
```

**After**:
```python
# Filter out model_id parameter to prevent "extraneous key" error from Bedrock API
optional_params.pop("model_id", None)

image_generation_config = {**image_generation_config, **optional_params}
```

#### 3. Comprehensive Test Coverage

Added 5 new test cases in `tests/image_gen_tests/test_bedrock_image_gen_unit_tests.py`:

- **ARN Format Support**: Tests inference profile ARN detection
- **Parameter Filtering**: Tests `model_id` parameter removal at both handler and transformer levels
- **Cross-Region Support**: Tests cross-region inference profile formats (`us.amazon.nova-canvas-v1:0`)
- **Backward Compatibility**: Regression test for existing `amazon.nova-canvas-v1:0` format

### Supported Model Formats

After this fix, all these formats work correctly:

✅ **Regular model**: `amazon.nova-canvas-v1:0` *(existing)*
✅ **ARN format**: `arn:aws:bedrock:eu-west-1:000000000000:application-inference-profile/a0a0a0a0a0a0` *(fixed)*
✅ **Cross-region format**: `us.amazon.nova-canvas-v1:0` *(fixed)*
✅ **Environment variable configs** with `model_id` parameter *(fixed)*

### Technical Approach

- **Pattern-Consistent**: Uses existing `get_bedrock_invoke_provider()` method that already handles ARN formats
- **Minimal Change**: Only 15 lines added across 2 files
- **Backward Compatible**: Maintains fallback logic for edge cases
- **Well-Tested**: 5 comprehensive test cases covering all scenarios

### Test Results

**New Tests (All Passing)**:
```
tests/image_gen_tests/test_bedrock_image_gen_unit_tests.py::test_get_request_body_nova_canvas_inference_profile_arn PASSED
tests/image_gen_tests/test_bedrock_image_gen_unit_tests.py::test_get_request_body_nova_canvas_with_model_id_param PASSED
tests/image_gen_tests/test_bedrock_image_gen_unit_tests.py::test_transform_request_body_nova_canvas_filter_model_id PASSED
tests/image_gen_tests/test_bedrock_image_gen_unit_tests.py::test_get_request_body_cross_region_inference_profile PASSED
tests/image_gen_tests/test_bedrock_image_gen_unit_tests.py::test_backward_compatibility_regular_nova_model PASSED
```

**Regression Testing**: 43/44 existing tests pass (1 failure due to missing boto3 dependency in test environment)

### Verification Steps

#### Automated Verification ✅
- [x] **Tests pass**: 5/5 new tests pass, 43/44 existing tests pass
- [x] **Linting passes**: `ruff check` reports no issues on modified files
- [x] **Type checking passes**: No type errors in implementation
- [x] **No regressions**: All existing Nova Canvas functionality preserved

#### Manual Verification Required
- [ ] Test with real AWS Bedrock inference profile ARNs
- [ ] Verify environment variable `model_id` configurations work end-to-end
- [ ] Performance testing under load with inference profiles

### Breaking Changes

**None** - This is a pure bug fix that enables existing documented functionality to work correctly.

### Migration Required

**None** - Existing configurations continue to work unchanged.

### Files Modified

1. **`litellm/llms/bedrock/image/image_handler.py`** (11 lines added)
   - Enhanced provider detection with ARN support

2. **`litellm/llms/bedrock/image/amazon_nova_canvas_transformation.py`** (4 lines added)
   - Added `model_id` parameter filtering

3. **`tests/image_gen_tests/test_bedrock_image_gen_unit_tests.py`** (92 lines added)
   - Comprehensive test coverage for all scenarios

### Impact Assessment

- **User Impact**: Resolves critical blocking issues for customers using inference profiles
- **Performance**: No performance impact - uses existing efficient methods
- **Compatibility**: Full backward compatibility maintained
- **Code Quality**: Follows existing LiteLLM patterns and conventions

timelfrink committed:
Add comprehensive tests for Bedrock inference profiles with Nova Canvas

- Test ARN format inference profile model detection
- Test model_id parameter filtering to prevent API errors
- Test cross-region inference profile format support
- Ensure backward compatibility with regular Nova Canvas models
- Cover both request body generation and parameter transformation

timelfrink committed:
Fix Bedrock inference profiles for Nova Canvas image generation

- Replace naive model.split('.')[0] with ARN-aware provider detection
- Use existing get_bedrock_invoke_provider() method for inference profile ARNs
- Filter model_id parameter to prevent 'extraneous key' API errors
- Support ARN format: arn:aws:bedrock:region:account:application-inference-profile/id
- Support cross-region format: us.amazon.nova-canvas-v1:0
- Maintain backward compatibility with regular amazon.nova-canvas-v1:0 format

krrishdholakia commented on litellm/llms/bedrock/image/amazon_nova_canvas_transformation.py:
```diff
@@ -67,6 +67,10 @@ def transform_request_body(
         image_generation_config = optional_params.pop("imageGenerationConfig", {})
+
+        # Filter out model_id parameter to prevent "extraneous key" error from Bedrock API
+        optional_params.pop("model_id", None)
```
is this how model id is handled on embeddings or chat completions?

timelfrink replied:
In image generation, model_id is extracted only to prevent API errors (not for usage like in chat/embeddings), since the image handler uses the main model parameter directly for the endpoint URL.

timelfrink committed:
Fix Nova Canvas model_id parameter handling for consistency

Follow the same pattern as chat completions and embeddings by extracting
model_id from optional_params into a variable, even though it's not used
in image generation. This maintains code consistency across Bedrock services.

timelfrink committed:
Document model_id parameter for Bedrock inference profiles with image generation

Add documentation for using AWS Bedrock Application Inference Profiles
with image generation APIs, specifically for Nova Canvas models.

The documentation includes:
- SDK examples showing model_id parameter usage
- Proxy configuration examples
- Follows the same pattern as chat completions inference profiles

timelfrink commented:
Confirmed locally:
<img width="1080" height="155" alt="image" src="https://github.com/user-attachments/assets/f4a84bcc-e4e1-4ec5-aab2-870858a00219" />
